File: VB/E1554.Module/ShowFilterDialogController.vb
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.SystemModule
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.Data.Filtering

Namespace E1554.Module
	Public Class ShowFilterDialogController
		Inherits WindowController

		Public Sub New()
			TargetWindowType = WindowType.Main
		End Sub

		Private showNavigationItemController As ShowNavigationItemController
		Private oldListView As ListView

		Protected Overrides Sub OnActivated()
			MyBase.OnActivated()
			showNavigationItemController = Frame.GetController(Of ShowNavigationItemController)()
			If showNavigationItemController IsNot Nothing Then
				AddHandler showNavigationItemController.CustomShowNavigationItem, AddressOf ShowNavigationItemController_CustomShowNavigationItem
			End If
		End Sub
		Private Sub ShowNavigationItemController_CustomShowNavigationItem(ByVal sender As Object, ByVal e As CustomShowNavigationItemEventArgs)
			Dim shortcut As ViewShortcut = TryCast(e.ActionArguments.SelectedChoiceActionItem.Data, ViewShortcut)
			If shortcut IsNot Nothing Then
				oldListView = TryCast(Application.ProcessShortcut(shortcut), ListView)
				If oldListView IsNot Nothing Then
					e.Handled = True
					Dim nonPersistentObjectSpace As NonPersistentObjectSpace = CType(Application.CreateObjectSpace(GetType(ViewFilterContainer)), NonPersistentObjectSpace)
					Dim persistentObjectSpace As IObjectSpace = Application.CreateObjectSpace(GetType(ViewFilterObject))
					nonPersistentObjectSpace.AdditionalObjectSpaces.Add(persistentObjectSpace)
					Dim newViewFilterContainer As ViewFilterContainer = nonPersistentObjectSpace.CreateObject(Of ViewFilterContainer)()
					newViewFilterContainer.ObjectType = oldListView.ObjectTypeInfo.Type
					newViewFilterContainer.Filter = GetFilterObject(persistentObjectSpace, DirectCast(oldListView.Model, IModelListViewAdditionalCriteria).AdditionalCriteria, newViewFilterContainer.ObjectType)
					Dim filterDetailView As DetailView = Application.CreateDetailView(nonPersistentObjectSpace, newViewFilterContainer)
					filterDetailView.Caption = String.Format("Filter for the {0} ListView", oldListView.Caption)
					filterDetailView.ViewEditMode = ViewEditMode.Edit
					e.ActionArguments.ShowViewParameters.CreatedView = filterDetailView
					e.ActionArguments.ShowViewParameters.TargetWindow = TargetWindow.NewModalWindow
					Dim dialogCotnroller As DialogController = Application.CreateController(Of DialogController)()
					AddHandler dialogCotnroller.Accepting, AddressOf dialogCotnroller_Accepting
					AddHandler dialogCotnroller.Cancelling, AddressOf dialogCotnroller_Cancelling
					AddHandler dialogCotnroller.ViewClosed, AddressOf dialogCotnroller_ViewClosed
					e.ActionArguments.ShowViewParameters.Controllers.Add(dialogCotnroller)
				End If
			End If
		End Sub
		Private Sub dialogCotnroller_Cancelling(ByVal sender As Object, ByVal e As EventArgs)
			oldListView.Dispose()
		End Sub
		Private Sub dialogCotnroller_Accepting(ByVal sender As Object, ByVal e As DialogControllerAcceptingEventArgs)
			Dim currentViewFilterContainer As ViewFilterContainer = CType(e.AcceptActionArgs.CurrentObject, ViewFilterContainer)
			Dim targetView As ListView = GetTargetView()
			DirectCast(targetView.Model, IModelListViewAdditionalCriteria).AdditionalCriteria = currentViewFilterContainer.Criteria
			targetView.CollectionSource.Criteria("ByViewFilterObject") = CriteriaEditorHelper.GetCriteriaOperator(currentViewFilterContainer.Criteria, currentViewFilterContainer.ObjectType, targetView.ObjectSpace)
			oldListView = Nothing
			ShowView(targetView, e.ShowViewParameters)
		End Sub
		Private Sub dialogCotnroller_ViewClosed(ByVal sender As Object, ByVal e As EventArgs)
			oldListView = Nothing
		End Sub
		Protected Overridable Sub ShowView(ByVal targetView As View, ByVal showViewParameters As ShowViewParameters)
			Window.SetView(targetView)
		End Sub
		Protected Overridable Function GetTargetView() As ListView
			Return oldListView
		End Function
		Private Function GetFilterObject(ByVal objectSpace As IObjectSpace, ByVal listViewCriteria As String, ByVal objectType As Type) As ViewFilterObject
			Dim criteria As CriteriaOperator = CriteriaOperator.Parse("Criteria = ? and ObjectType = ?", listViewCriteria, objectType)
			Dim filterObject As ViewFilterObject = objectSpace.FindObject(Of ViewFilterObject)(criteria)
			If filterObject Is Nothing Then
				filterObject = objectSpace.CreateObject(Of ViewFilterObject)()
				filterObject.ObjectType = objectType
				filterObject.Criteria = listViewCriteria
				filterObject.FilterName = "Default"
			End If
			Return filterObject
		End Function
		Protected Overrides Sub OnDeactivated()
			MyBase.OnDeactivated()
			If showNavigationItemController IsNot Nothing Then
				RemoveHandler showNavigationItemController.CustomShowNavigationItem, AddressOf ShowNavigationItemController_CustomShowNavigationItem
				showNavigationItemController = Nothing
			End If
			oldListView = Nothing
		End Sub
	End Class
End Namespace

File: CS/E1554.Module/ShowFilterDialogController.cs
using System;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.SystemModule;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.Editors;
using DevExpress.Data.Filtering;

namespace E1554.Module {
    public class ShowFilterDialogController : WindowController {
        public ShowFilterDialogController() {
            TargetWindowType = WindowType.Main;
        }

        ShowNavigationItemController showNavigationItemController;
        ListView oldListView;
        ChoiceActionItem selectedChoiceActionItem;
        bool showFromController = false;

        protected override void OnActivated() {
            base.OnActivated();
            showNavigationItemController = Frame.GetController<ShowNavigationItemController>();
            if (showNavigationItemController != null) {
                showNavigationItemController.CustomShowNavigationItem += ShowNavigationItemController_CustomShowNavigationItem;
            }
        }
        private void ShowNavigationItemAction_Executed(object sender, ActionBaseEventArgs e) {
            showNavigationItemController.ShowNavigationItemAction.Executed -= ShowNavigationItemAction_Executed;
            if(showFromController) {
                e.ShowViewParameters.CreatedView = oldListView;
            }
        }
        private void ShowNavigationItemController_CustomShowNavigationItem(object sender, CustomShowNavigationItemEventArgs e) {
            if (showFromController) {
                return;
            }
            ViewShortcut shortcut = e.ActionArguments.SelectedChoiceActionItem.Data as ViewShortcut;
            if(shortcut != null) {
                oldListView = Application.ProcessShortcut(shortcut) as ListView;
            }
            if(oldListView != null) {
                e.Handled = true;
                selectedChoiceActionItem = e.ActionArguments.SelectedChoiceActionItem;
                NonPersistentObjectSpace nonPersistentObjectSpace = (NonPersistentObjectSpace)Application.CreateObjectSpace(typeof(ViewFilterContainer));
                IObjectSpace persistentObjectSpace = Application.CreateObjectSpace(typeof(ViewFilterObject));
                nonPersistentObjectSpace.AdditionalObjectSpaces.Add(persistentObjectSpace);
                ViewFilterContainer newViewFilterContainer = nonPersistentObjectSpace.CreateObject<ViewFilterContainer>();
                newViewFilterContainer.ObjectType = oldListView.ObjectTypeInfo.Type;
                newViewFilterContainer.Filter = GetFilterObject(persistentObjectSpace, ((IModelListViewAdditionalCriteria)oldListView.Model).AdditionalCriteria, newViewFilterContainer.ObjectType);
                DetailView filterDetailView = Application.CreateDetailView(nonPersistentObjectSpace, newViewFilterContainer);
                filterDetailView.Caption = String.Format("Filter for the {0} ListView", oldListView.Caption);
                filterDetailView.ViewEditMode = ViewEditMode.Edit;
                e.ActionArguments.ShowViewParameters.CreatedView = filterDetailView;
                e.ActionArguments.ShowViewParameters.TargetWindow = TargetWindow.NewModalWindow;
                DialogController dialogCotnroller = Application.CreateController<DialogController>();
                dialogCotnroller.Accepting += new EventHandler<DialogControllerAcceptingEventArgs>(dialogCotnroller_Accepting);
                dialogCotnroller.ViewClosed += dialogCotnroller_ViewClosed;
                e.ActionArguments.ShowViewParameters.Controllers.Add(dialogCotnroller);
            }
        }
        void dialogCotnroller_Accepting(object sender, DialogControllerAcceptingEventArgs e) {
            ViewFilterContainer currentViewFilterContainer = (ViewFilterContainer)e.AcceptActionArgs.CurrentObject;
            ListView targetView = GetTargetView();
            ((IModelListViewAdditionalCriteria)targetView.Model).AdditionalCriteria = currentViewFilterContainer.Criteria;
            targetView.CollectionSource.Criteria["ByViewFilterObject"] = CriteriaEditorHelper.GetCriteriaOperator(currentViewFilterContainer.Criteria, currentViewFilterContainer.ObjectType, targetView.ObjectSpace);
            showNavigationItemController.ShowNavigationItemAction.Executed += ShowNavigationItemAction_Executed;
            try {
                showFromController = true;
                showNavigationItemController.ShowNavigationItemAction.DoExecute(selectedChoiceActionItem);
            }
            finally {
                showFromController = false;
                oldListView = null;
                selectedChoiceActionItem = null;
            }
        }

        protected virtual ListView GetTargetView() {
            return oldListView;
        }

        void dialogCotnroller_ViewClosed(object sender, EventArgs e) {
            oldListView = null;
            selectedChoiceActionItem = null;
        }

        private ViewFilterObject GetFilterObject(IObjectSpace objectSpace, string listViewCriteria, Type objectType) {
            CriteriaOperator criteria = CriteriaOperator.Parse("Criteria = ? and ObjectType = ?", listViewCriteria, objectType);
            ViewFilterObject filterObject = objectSpace.FindObject<ViewFilterObject>(criteria);
            if (filterObject == null) {
                filterObject = objectSpace.CreateObject<ViewFilterObject>();
                filterObject.ObjectType = objectType;
                filterObject.Criteria = listViewCriteria;
                filterObject.FilterName = "Default";
            }
            return filterObject;
        }
        protected override void OnDeactivated() {
            base.OnDeactivated();
            if (showNavigationItemController != null) {
                showNavigationItemController.CustomShowNavigationItem -= ShowNavigationItemController_CustomShowNavigationItem;
                showNavigationItemController = null;
            }
            oldListView = null;
            selectedChoiceActionItem = null;
        }
    }
}
